Encryption.md

```diff
@@ -75,55 +75,76 @@ written to the output stream.
 write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
 with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
 the bulk of the data faster, while still verifying the metadata integrity and making sure 
-the file had not been replaced with a wrong version. However, tampering with the page data 
+the file has not been replaced with a wrong version. However, tampering with the page data 
 might go unnoticed. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
-
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, 
+since a single re-use of an IV breaks the encryption. An IV in these two modes is often called a 
+nonce (“number used once”). Parquet encryption uses the RBG-based (random bit generator) IV 
+construction as defined in the NIST SP 800-38D document for the GCM ciphers (section 8.2.2). 
 
 
 ### AES_GCM_V1
-All modules are encrypted with the AES-GCM cipher. The authentication tags (16 bytes) are 
-written after each ciphertext. The IVs (12 bytes) are written before each ciphertext.
+All modules are encrypted by an AES-GCM cipher, without padding. The authentication tags (16 bytes) are 
+written after each ciphertext. The IVs are written before each ciphertext. The IV length is 12 
+bytes (96 bits), which is generally considered to be the optimal size of GCM IVs.  The NIST 
+specification requires a “random field” to fill all 12 bytes and an arbitrary “free field” to be empty for 
+this IV length (see section 8.2.2).
 
 ### AES_GCM_CTR_V1
-Thrift modules are encrypted with the AES-GCM cipher, as described above. 
-The pages are encrypted with AES-CTR, where the IVs (16 bytes) are written before each 
-ciphertext.
+Thrift modules are encrypted with AES-GCM, as described above. The pages are encrypted by an AES-CTR
+cipher without padding. The nonces are written before each ciphertext, with the nonce length of 16 bytes 
+(128 bits). At least 12 bytes out of 16 are constructed with an RBG.
+
+```c
+struct AesGcmV1 {
+  /** Retrieval metadata of AAD used for encryption of pages and structures **/
+  1: optional binary aad_metadata
+}
+
+struct AesGcmCtrV1 {
+  /** Retrieval metadata of AAD used for encryption of structures **/
+  1: optional binary aad_metadata
+}
+
+union EncryptionAlgorithm {
+  1: AesGcmV1 AES_GCM_V1
+  2: AesGcmCtrV1 AES_GCM_CTR_V1
+}
+```
 
+The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
+be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
```

Comment:
This basically says "There is an AAD field to store the AAD." You could briefly explain what the purpose of that field is. If my understanding is correct, that would be something like this:

```suggestion
be used to store any user-defined Additional Authenticated Data. The AAD itself is not
encrypted, but it is attached to the ciphertext in a tamperproof manner. The maximal allowed
```